lnbits/wallets/nwc.py

```diff
@@ -0,0 +1,881 @@
+
+    async def _on_message(self, ws, message: str):
+        """
+        Handle incoming messages from the relay.
+        """
+        try:
+            msg = json.loads(message)
+            if msg[0] == "OK":  # Event status message
+                await self._on_ok_message(msg)
+            elif msg[0] == "EVENT":  # Event message
+                await self._on_event_message(msg)
+            elif msg[0] == "EOSE":
+                # Do nothing. No need to handle this message type for NWC
+                pass
+            elif msg[0] == "CLOSED":
+                # Subscription was closed remotely.
+                await self._on_closed_message(msg)
+            elif msg[0] == "NOTICE":
+                # A message from the relay, mostly useless, but we log it anyway
+                logger.info("Notice from relay " + self.relay + ": " + str(msg[1]))
+            else:
+                raise Exception("Unknown message type")
+        except Exception as e:
+            logger.error("Error parsing event: " + str(e))
+
+    async def _connect_to_relay(self):
+        """
+        Initiate websocket connection to the relay.
+        """
+        logger.debug("Connecting to NWC relay " + self.relay)
+        while (
+            not self._is_shutting_down()
+        ):  # Reconnect until the wallet is shutting down
+            logger.debug("Creating new connection...")
+            try:
+                async with ws_connect(self.relay) as ws:
+                    self.ws = ws
+                    self.connected = True
+                    while (
+                        not self._is_shutting_down()
+                    ):  # receive messages until the wallet is shutting down
+                        try:
+                            reply = await ws.recv()
+                            reply_str = ""
+                            if isinstance(reply, bytes):
+                                reply_str = reply.decode("utf-8")
+                            else:
+                                reply_str = reply
+                            await self._on_message(ws, reply_str)
+                        except Exception as e:
+                            logger.debug("Error receiving message: " + str(e))
+                            break
+                logger.debug("Connection to NWC relay closed")
+            except Exception as e:
+                logger.error("Error connecting to NWC relay: " + str(e))
+            # the connection was closed, so we set the connected flag to False
+            # this will make the methods calling _wait_for_connection()
+            # to wait until the connection is re-established
+            self.connected = False
+            if not self._is_shutting_down():
+                # Wait some time before reconnecting
+                logger.debug("Reconnecting to NWC relay in 5 seconds...")
+                await asyncio.sleep(5)
+
+    def _encrypt_content(self, content: str) -> str:
```

Comment:
Regarding this and the requested changes for :  _decrypt_content, _verify_event and _sign_event.

Usually when i write this kind of code i try to logically confine it as much as possible and to avoid passing around parameters that never change, this is to prevent future oversights from breaking core/security functionalities.

If i were to make it an helper i would want at least to have secp256k1.PrivateKey and secp256k1.PublicKey as parameters, to reduce possible mistakes, but this would require more conversions back and forth, so having it as a class method, imo, is a shortcut that is a good compromise.

Please consider this point of view, I know the downside is that it hurts re-usability (and that i might be paranoid) so if you still want the change i can do it.

Reply:
> Usually when i write this kind of code i try to logically confine it as much as possible and to avoid passing around parameters that never change, this is to prevent future oversights from breaking core/security functionalities.

That is my point, there are no parameters and `self` is not required. These functions can be moved outside the class on the bottom of the file.

> If i were to make it an helper i would want at least to have secp256k1.PrivateKey and secp256k1.PublicKey as parameters

No need for this.

Reply:
I don't understand what you mean, these function do have parameters and use self in the way i've implemented.
My point is that the way you want them implemented makes for low quality helpers that are prone to misuse and errors (eg. passing the wrong key)

Reply:
> My point is that the way you want them implemented makes for low quality helpers that are prone to misuse and errors (eg. passing the wrong key)

- the function name and parameters must reflect the functionality. For example `_encrypt_content(...)` needs both a `content` to be encrypted and a `secret` to encrypt with
- the function needs to be re-usable, if the `secret` is not a parameter then it means it is **hardcoded** inside the function. Therefore this function cannot be reused with other secrets
- `self` should not be present in these helper functions (they do not need access to the parent class)
- `passing the wrong key` - following this logic we should not use parameters for functions because we can always pass the wrong values

Reply:
I will make the change you asked anyway, since at the end of the day you are the maintainer/designer here, but i strongly disagree with your points.

Especially with:

> if the secret is not a parameter then it means it is hardcoded inside the function

It is not hardcoded, it is passed in the constructor and stored as class field. When something is hardcoded it means it can never change unless you go and change it in the source. You could call this immutable (it would be if python had this possibility) or whatever, but not hardcoded.


> the function name and parameters must reflect the functionality. For example _encrypt_content(...) needs both a content to be encrypted and a secret to encrypt with
> passing the wrong key - following this logic we should not use parameters for functions because we can always pass the wrong values

It is perfectly ok to have methods that are not reusable (private final), that are used internally by a class, and these encryption/decryption methods are exactly like that.
They are not general purpose and should not be treated as such, they do only one thing: encrypt and decrypt the content of a message for **the instantiated** nwc connection, you should never pass a different secret there (ever).

What you are proposing is an antipattern in oop, methods should always have the minimum numbers of arguments possible and, if they aren't general purpose, they should go in the class where they are used. The preferable way to reuse code in oop is by extending classes, not by borrowing internal helper functions.

What you propose seems fine because python doesn't have access modifiers, so everything is public, but try to make a private global scoped helper in another language.

Reply:
Done :+1:

Reply:
Thank you!
Not trying to give you a hard time here btw, you did a great job with this PR!

Reply:
Thank you. No problem, a bit of debate doesn’t bother me at all. I appreciate our disagreement on the matter and apologies if I came across too strongly. It is just a styling choice at the end of the day.